src/2d/shallow/bc2amr.f90

```diff
@@ -0,0 +1,280 @@
+                do i = ibeg, nrow
+                    do j = 1, ncol
+                        aux(:, i, j) = aux(:, ibeg - 1, j)
+                        val(:, i, j) = val(:, ibeg - 1, j)
+                    end do
+                end do
+
+            case(2) ! Periodic boundary condition
+                continue
+
+            case(3) ! Wall boundary conditions
+                do i=ibeg, nrow
+                    do j = 1, ncol
+                        aux(:, i, j) = aux(:, 2 * ibeg - 1 - i, j)
+                        val(:, i, j) = val(:, 2 * ibeg - 1 - i, j)
+                    end do
+                end do
+                ! negate the normal velocity:
+                do i = ibeg, nrow
+                    do j = 1, ncol
+                        val(2, i, j) = -val(2, i, j)
+                    end do
+                end do
+
+            case default
+                print *, "Invalid boundary condition requested."
+                stop
+
+        end select
+    end if
+
+    !-------------------------------------------------------
+    ! Bottom boundary:
+    !-------------------------------------------------------
+    if (ylo_patch < ylower - hymarg) then
+
+        ! number of grid cells lying outside physical domain:
+        nyb = int((ylower + hymarg - ylo_patch) / hy)
+
+        select case(mthbc(3))
+            case(0) ! User defined boundary condition
+                ! Replace this code with a user defined boundary condition
+                stop "A user defined boundary condition was not provided."
+            
+            case(1) ! Zero-order extrapolation
+                do j = 1, nyb
+                    do i = 1, nrow
+                        aux(:,i,j) = aux(:, i, nyb + 1)
+                        val(:,i,j) = val(:, i, nyb + 1)
+                    end do
+                end do
+
+            case(2) ! Periodic boundary condition
+                continue
+
+            case(3) ! Wall boundary conditions
+                do j = 1, nyb
+                    do i = 1, nrow
+                        aux(:,i,j) = aux(:, i, 2 * nyb + 1 - j)
+                        val(:,i,j) = val(:, i, 2 * nyb + 1 - j)
+                    end do
+                end do
+                ! negate the normal velocity:
+                do j = 1, nyb
+                    do i = 1, nrow
+                        val(3,i,j) = -val(3, i, j)
+                    end do
+                end do
+
+            case default
+                print *, "Invalid boundary condition requested."
+                stop
+
+        end select
+    end if
+
+    !-------------------------------------------------------
+    ! Top boundary:
+    !-------------------------------------------------------
+    if (yhi_patch > yupper + hymarg) then
+
+        ! number of grid cells lying outside physical domain:
+        nyt = int((yhi_patch - yupper + hymarg) / hy)
+        jbeg = max(ncol - nyt + 1, 1)
+
+        select case(mthbc(3))
```

Comment:
I think this should check `mthbc(4)` not 3.

Reply:
Yeah, that would probably be good wouldn't it?